src/contribute/portability.md

```diff
@@ -0,0 +1,164 @@
+user-customized execution environments ("custom containers").
+
+The portability API consists of a set of smaller contracts that
+isolate SDKs and runners for job submission, management and
+execution. These contracts use protobufs and gRPC for broad language
+support.
+
+ * **Job submission and management**: The _Runner API_ defines a
+   language-neutral pipeline representation with transformations
+   specifying the execution environment as a docker container
+   image. The latter both allows the execution side to set up the
+   right environment as well as opens the door for custom containers
+   and cross-environment pipelines. The _Job API_ allows pipeline
+   execution and configuration to be managed uniformly.
+
+ * **Job execution**: The _Fn API_ defines execution-time binary
+   contract between the SDK harness and the runner for how execution
+   tasks are managed and how data is transferred. In addition, the
+   runner needs to handle progress and monitoring in an efficient and
+   language-neutral way. SDK harness initialization relies on the
+   _Provision_ and _Artifact APIs_ for obtaining staged files,
+   pipeline options and environment information. Docker provides
+   isolation between the runner and SDK/user environments to the
+   benefit of both as defined by the _container contract_. The
+   containerization of the SDK gives it (and the user, unless the SDK
+   is closed) full control over its own environment without risk of
+   dependency conflicts. The runner has significant freedom regarding
+   how it manages the SDK harness containers.
+
+The goal is that all (non-direct) runners and SDKs eventually support
+the portability API, perhaps exclusively.
+
+## Design
+
+The [model protos](https://github.com/apache/beam/tree/master/model)
+contain all aspects of the portability API and is the truth on the
+ground. The proto definitions supercede any design documents. The main
+design documents are the following:
+
+ * [Runner API](https://s.apache.org/beam-runner-api). Pipeline
+   representation and discussion on primitive/composite transforms and
+   optimizations.
+   
+ * [Job API](https://s.apache.org/beam-job-api). Job submission and
+   management protocol.
+
+ * [Fn API](https://s.apache.org/beam-fn-api). Execution-side control
+   and data protocols and overview.
+
+ * [Container
+   contract](https://s.apache.org/beam-fn-api-container-contract).
+   Execution-side docker container invocation and provisioning
+   protocols. See
+   [CONTAINERS.md](https://github.com/apache/beam/blob/master/sdks/CONTAINERS.md)
+   for how to build container images.
+
+In discussion:
+
+ * [Cross
+   language](https://s.apache.org/beam-mixed-language-pipelines). Options
+   and tradeoffs for how to handle various kinds of
+   multi-language/multi-SDK pipelines.
+
+## Development
+
+The portability framework is a substantial effort that touches every
+Beam component. Other than the sheer magnitude, a major challenge is
```

Comment:
suggestion for slight reword of the first part of this sentence:
In addition to the sheer magnitude, a major challenge is engineering an interop layer that does not significantly compromise performance due ...